External Service Button

rmkraus opened the pull request:
Added an HTTPS button to the VM's action panel. Clicking this
button will open a broser tab on port 443 of the VM. Currently, the
button will be shown as long as the VM is running.

~~This PR is to address #1155 and is currently a work in progress. I am looking for feedback on how to make this code acceptable for merging.~~

Fixes #1155 
Fixes #1163

Thanks!

ovirt-infra commented:
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically. ( keep in mind this list might be overwritten if the job XML is refreshed, for permanent whitelisting, please follow #3 option )
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

rmkraus commented:
This is a screenshot of the UI change. 

<img width="539" alt="Screen Shot 2020-01-22 at 3 29 19 AM" src="https://user-images.githubusercontent.com/4956442/72877664-6b9f9380-3cc7-11ea-9e34-e5781514a30d.png">

michalskrivanek commented on src/ovirtapi/transform.js:
```diff
@@ -117,6 +118,12 @@ const VM = {
 
       fqdn: vm['fqdn'],
 
+      service: {
```
it would be better to encapsulate that in a single property, like external_service "name/proto/port" and trivially parse it here into those three components

rmkraus replied:
:+1: 
I have made this change.

michalskrivanek commented on src/components/VmActions/index.js:
```diff
@@ -249,16 +250,22 @@ class VmActions extends React.Component {
-      {
+    ]
+
+    if (canService(vm.get('service'))) {
```
would probably be better to keep the transform.js generic and only expose custom properties, and this canService(or canExternalService) would look up the correct property and parse that x/y/z value

rmkraus replied:
:+1:

This has been done in the newest commit.

michalskrivanek commented on src/components/VmActions/index.js:
```diff
@@ -249,16 +250,22 @@ class VmActions extends React.Component {
         actionDisabled: isPool || !canConsole(status),
-        shortTitle: msg.https(),
+        shortTitle: vm.get('service').get('name'),
         tooltip: msg.httpsService(),
```
maybe add the name to the tooltip too?

rmkraus replied:
I'm not sure. I'm happy to do it, but it feels a bit like using the word you're defining in the definition. What do you think?

rmkraus commented:
I've made most of the requested changes. I'm open to conversation on the Tooltip, I don't have any strong feelings on it. We can squash when merging. I think that would be safer than trusting my git-fu! ;-)

This newest version uses a single custom property called `external_service`. The value of this property is a bar delineated list of NAME|PROTOCOL|PORT. Additional services can be defined by separating them with a comma. So:
Service 1|HTTPS|443,Service 2|HTTPS|8080,Service 3|HTTP|80

For every properly formatted service record in the string, a button will be created on the frontend. The button will only be usable if a FQDN is available, the VM canConsole, and it is not a Pool card.

I thought briefly about having all the service buttons in a drop down. However, these services will likely be unrelated things (Ex Cockpit, Jupyter Lab, and PhpMyAdmin) and putting them in one drop down may not be the best UX.

rmkraus commented on src/components/VmActions/index.js:
```diff
@@ -80,6 +81,7 @@ const VmDropdownActions = ({ actions, id }) => {
             confirmation={action.confirmation}
             shortTitle={action.shortTitle}
             icon={action.icon}
+            onClick={action.onClick}
```
I want to draw attention to this change. I think this was a bug, but I could also be missing something.

The onClick action wasn't being passed to the VM Drop Down Action buttons on the VM card view. Only the primary action kept its onClick property. This never caused an issue because all the other actions use a confirmation dialog which was being passed.

michalskrivanek replied:
no idea. possibly a bug indeed. @sjd78 ?

rmkraus replied:
I'm not sure I know what you mean on this one. Could you provide a bit more detail? 

Right now, MenuItemAction does this for onClick:

```javascript
onClick={() => {                                                         
  onClick && onClick()                                                   
  document.dispatchEvent(new MouseEvent('click'))                        
}
```

sjd78 replied:
Sure.  Here is the action definition for RDP console, which has the `onClick` definition:
https://github.com/oVirt/ovirt-web-ui/blob/09be1f1b36f5972e0efa6c41ca3128053872db58/src/components/VmActions/index.js#L162-L169

The onClick's `e.preventDefault()` will fail since, with the `MenuItemAction` `onClick` implementation you listed, nothing is passed in and e will be `undefined`.  The fix would be the have the `MenuItemAction` `onClick` implementation be:

```jsx
    <MenuItem
      onClick={e => {
        onClick && onClick(e)
        document.dispatchEvent(new MouseEvent('click'))
      }}
      {...rest}
    >
      <span>{shortTitle}</span> {icon}
    </MenuItem>
```

or to use the ... operators to handle all potential function arguments:
```jsx
    <MenuItem
      onClick={(...args) => {
        onClick && onClick(...args)
        document.dispatchEvent(new MouseEvent('click'))
      }}
      {...rest}
    >
      <span>{shortTitle}</span> {icon}
    </MenuItem>
```

Hope that helps.

You can either add that to this PR or spin off a new issue + PR just to handle the "Console RDP button doesn't work".

rmkraus replied:
Thanks so much for taking the time to explain this. I think I got it fixed.

michalskrivanek commented:
> I've made most of the requested changes. I'm open to conversation on the Tooltip, I don't have any strong feelings on it. We can squash when merging. I think that would be safer than trusting my git-fu! ;-)

ok,np

> This newest version uses a single custom property called `external_service`. The value of this property is a bar delineated list of NAME|PROTOCOL|PORT. Additional services can be defined by separating them with a comma. So:
> Service 1|HTTPS|443,Service 2|HTTPS|8080,Service 3|HTTP|80

thanks, that looks good. I think it would be nicer with "/" (we use that syntax in osinfo) but it's fine as is:) 
 
> For every properly formatted service record in the string, a button will be created on the frontend. The button will only be usable if a FQDN is available, the VM canConsole, and it is not a Pool card.

does it have to be connected to canConsole? it may have a bit too many states...
export function canConsole (state) {
  return ['up', 'powering_up', 'powering_down', 'paused', 'migrating', 'reboot_in_progress', 'saving_state'].includes(state)
}

it may make sense to introduce a new one or use e.g. canSuspend(yeah, the name is inconvenient), but I would limit the states to something when the VM is running "normally". if that's a guest service it should be only Up, maybe Migrating too. 
Button(s) should be just grayed out instead of creating/hiding them all the time, ideally.

> I thought briefly about having all the service buttons in a drop down. However, these services will likely be unrelated things (Ex Cockpit, Jupyter Lab, and PhpMyAdmin) and putting them in one drop down may not be the best UX.

it's fine as is. thanks

michalskrivanek commented:
> it may make sense to introduce a new one or use e.g. canSuspend(yeah, the name is inconvenient), but I would limit the states to something when the VM is running "normally". if that's a guest service it should be only Up, maybe Migrating too.

that would be canRestart which we can use right away

rmkraus commented:
> > This newest version uses a single custom property called `external_service`. The value of this property is a bar delineated list of NAME|PROTOCOL|PORT. Additional services can be defined by separating them with a comma. So:
> > Service 1|HTTPS|443,Service 2|HTTPS|8080,Service 3|HTTP|80
> 
> thanks, that looks good. I think it would be nicer with "/" (we use that syntax in osinfo) but it's fine as is:)

Changed to be more inline with what is currently done. Having multiple standards doesn't feel good.

> 
> > For every properly formatted service record in the string, a button will be created on the frontend. The button will only be usable if a FQDN is available, the VM canConsole, and it is not a Pool card.
> 
> does it have to be connected to canConsole? it may have a bit too many states...
> export function canConsole (state) {
> return ['up', 'powering_up', 'powering_down', 'paused', 'migrating', 'reboot_in_progress', 'saving_state'].includes(state)
> }
> 
> it may make sense to introduce a new one or use e.g. canSuspend(yeah, the name is inconvenient), but I would limit the states to something when the VM is running "normally". if that's a guest service it should be only Up, maybe Migrating too.
> Button(s) should be just grayed out instead of creating/hiding them all the time, ideally.
> 

Makes sense. Changed to canRestart as requested.

sjd78 commented:
@rmkraus - Also, please squash your commits.  `git rebase -i` and set most to `s`.  This size change really only needs 1 commit.

rmkraus commented on src/components/VmActions/index.js:
```diff
@@ -250,6 +252,30 @@ class VmActions extends React.Component {
+      const services = servicesRaw.get('value').split(',').map(svc => svc.split('/')).filter(svc => svc.length === 3)
+
+      var idx = 0
+      for (const svc of services) {
```
👍 

I'm going to go with the last example. I like the deconstruction, but I also like the filter when parsing the service entries. It prevents an if statement later.

rmkraus commented on src/components/VmActions/index.js:
```diff
@@ -250,6 +252,30 @@ class VmActions extends React.Component {
+          priority: 0,
+          actionDisabled: isPool || !canExternalService(status, fqdn),
+          shortTitle: svc[0],
+          tooltip: msg.serviceTip(),
```
👍

rmkraus commented on src/intl/messages.js:
```diff
@@ -522,6 +522,7 @@ export const messages: { [messageId: string]: MessageType } = {
   secondsShort: 's',
   sendShortcutKey: 'Send Key',
   sendCtrlAltDel: 'Ctrl+Alt+Del',
+  serviceTip: 'Launch the VM\'s Service',
```
👍

rmkraus commented on src/components/VmActions/index.js:
```diff
@@ -250,6 +252,30 @@ class VmActions extends React.Component {
+      // multiple services can be defined by chaining them with a comma
+      const services = servicesRaw.get('value').split(',').map(svc => svc.split('/')).filter(svc => svc.length === 3)
+
+      var idx = 0
```
👍

rmkraus commented on src/components/VmActions/index.js:
```diff
@@ -250,6 +252,30 @@ class VmActions extends React.Component {
+    // external_service custom VM param.
+    const fqdn = vm.get('fqdn')
+    const servicesRaw = vm.get('customProperties').find(prop => prop.get('name') === 'external_service') || new Map()
+    if (servicesRaw.get('value')) {
```
👍 

Definitely more readable that way! Thanks.

rmkraus commented:
@sjd78 

I implemented all of your suggestions and squashed my commits. 👍 

I'm not sure what other change you wanted for the VM Actions onClick property.

rmkraus commented:
@sjd78 I put in the fix for the onClick event. Thanks for pointing all that out. I completely missed that! I think this PR should be good with all the requested changes and fixes at this point.

sjd78 commented:
> @sjd78 I put in the fix for the onClick event. Thanks for pointing all that out. I completely missed that! I think this PR should be good with all the requested changes and fixes at this point.

I created #1163 to capture the problem.  Now this PR gets to fix 2 issues.  Thanks!

rmkraus commented:
👍 

I'm ready to merge when you are.

sjd78 commented:
ci test please
